pkg/sparrow/controller.go

```diff
@@ -0,0 +1,270 @@
+
+	for _, c := range cc.checks.Iter() {
+		cErr := cc.UnregisterCheck(ctx, c)
+		if cErr != nil {
+			log.Error("Failed to unregister check while shutting down", "error", cErr)
+			err = errors.Join(err, cErr)
+		}
+	}
+	cc.done <- struct{}{}
+	close(cc.done)
+	return err
+}
+
+// Reconcile reconciles the checks.
+// It registers new checks, updates existing checks and unregisters checks not in the new config.
+func (cc *ChecksController) Reconcile(ctx context.Context, cfg runtime.Config) {
+	log := logger.FromContext(ctx)
+
+	newChecks, err := factory.NewChecksFromConfig(cfg)
+	if err != nil {
+		log.ErrorContext(ctx, "Failed to create checks from config", "error", err)
+		return
+	}
+
+	// Update existing checks and create a list of checks to unregister
+	var unregList []checks.Check
+	for _, c := range cc.checks.Iter() {
+		conf := cfg.For(c.Name())
+		if conf != nil {
+			err = c.SetConfig(conf)
+			if err != nil {
+				log.ErrorContext(ctx, "Failed to set config for check", "check", c.Name(), "error", err)
+			}
+			delete(newChecks, c.Name())
+		} else {
+			unregList = append(unregList, c)
+		}
+	}
+
+	// Unregister checks not in the new config
+	for _, c := range unregList {
+		err = cc.UnregisterCheck(ctx, c)
+		if err != nil {
+			log.ErrorContext(ctx, "Failed to unregister check", "check", c.Name(), "error", err)
+		}
+	}
+
+	// Register new checks
+	for _, c := range newChecks {
+		err = cc.RegisterCheck(ctx, c)
+		if err != nil {
+			log.ErrorContext(ctx, "Failed to register check", "check", c.Name(), "error", err)
+		}
+	}
+}
+
+// RegisterCheck registers a new check.
+func (cc *ChecksController) RegisterCheck(ctx context.Context, check checks.Check) error {
+	log := logger.FromContext(ctx).With("check", check.Name())
+
+	// Add prometheus collectors of check to registry
+	for _, collector := range check.GetMetricCollectors() {
+		if err := cc.metrics.GetRegistry().Register(collector); err != nil {
+			log.ErrorContext(ctx, "Could not add metrics collector to registry")
+		}
+	}
+
+	go func() {
+		err := cc.runCheck(ctx, check)
+		if err != nil {
+			log.ErrorContext(ctx, "Failed to start check", "error", err)
+			cc.cErr <- err
+		}
+	}()
+
+	cc.checks.Add(check)
+	return nil
+}
+
+// UnregisterCheck unregisters a check.
+func (cc *ChecksController) UnregisterCheck(ctx context.Context, check checks.Check) error {
+	log := logger.FromContext(ctx).With("check", check.Name())
+
+	// Remove prometheus collectors of check from registry
+	for _, metricsCollector := range check.GetMetricCollectors() {
+		if !cc.metrics.GetRegistry().Unregister(metricsCollector) {
+			log.ErrorContext(ctx, "Could not remove metrics collector from registry")
+		}
+	}
+
+	err := check.Shutdown(ctx)
+	if err != nil {
+		log.ErrorContext(ctx, "Failed to shutdown check", "error", err)
+		return err
+	}
+
+	for i, existingCheck := range cc.checks.Iter() {
+		if existingCheck.Name() == check.Name() {
+			cc.checks.Delete(i)
+			break
+		}
+	}
+
```

Comment:
I think it would make sense to refactor this to a `removeCheck(c check)` for the controller. We should try and hide away how the state is implemented (the current implementation is a slice of `Checks` but we shouldn't care about it). In case we want to switch implementations again (choose a map), such functions guard us from having to change too much of our codebase.

I strongly think this isn't a YAGNI feature but a needed part of the behavior of the ChecksController. As the `RegisterCheck` and `UnregisterCheck` dictate the controller's business logic, similar functions should handle its internal state (like the proposed `removeCheck`).